### DISCUSSION TITLE
OTP applications names

### DISCUSSION THREAD

**Discussion by elpitero** [keycloak/keycloak#17486] [Q&A]
Hello,
I noticed that when I select _Mobile application_ as an authentication method, i see the following as app names:

- "totpAppGoogleName" instead of  "Google authenticator"
- "totpAppFreeOTPName" instead of "FreeOTP"

To list names i use 
` <#list totp.supportedApplications as app> <li>${app}</li> </#list>`
I looked into the code here on github and i saw that in OtpApplicationProvider classes getName() returns this new string instead of old application name. Is there any specific reason for getName to return this string?

**Comment by qudrat220** [comment]
HBI

**Comment by davince72** [comment]
I have same issue, it seems it's not resolving the messages from the localization anymore.  It does for `totpStep2`, but not for `totpAppFreeOTPName` although it's defined in message file.
We now see (partial example):
```
Install one of the following applications on your mobile:

      totpAppFreeOTPName
      totpAppGoogleName
      totpAppMicrosoftAuthenticatorName
Open the application and scan the QR code.
```
ANy idea how to fix? (keycloak 23.11)

**Comment by danielFesenmeyer** [comment]
One way (until fixed in Keycloak) is to provide a custom theme, which specifies the missing translations in the messages*.properties files. This makes sense when you already provide a custom theme.

The easiest way (without "coding") is to define realm localization texts for the missing keys: https://www.keycloak.org/docs/latest/server_admin/#:~:text=realm%2Dspecific%20texts%20on%20the%20Localization%20tab

**Comment by elpitero** [comment]
Hi, I forgot about this thread but the solution i found was to wrap {app} with msg:
`
<#list totp.supportedApplications as app>
    <li>${msg(app)}</li>
</#list>
 `
Translations are already defined in messages_en.properties file. (After keycloak upgraded to version 20.0.0 getName method doesn't return name, but rather properties key)
